### PR TITLE
HCAP-1014: HA CSV hiring report by health region

### PR DESCRIPTION
**Pull request by alva-fresh**
https://freshworks.atlassian.net/browse/HCAP-1014

❔ **What's included:** 🎉 
- general refactoring to match current designs 🎨 
- generating report based on health region for both HA and MoH
- downloading report

https://www.figma.com/file/35c5C0DyTEOxJ4N4DIaADL/HCAP-Design

**HA:**
Screen Shot 2022-05-05 at 6 24 57 PM
Screen Shot 2022-05-06 at 3 19 38 PM

### CODE REVIEW

File: client/src/pages/private/SiteTable.js [hunk 1]
```diff
@@ -23,14 +43,99 @@ const columns = [
...
+            operatorPhone: '',
+            operatorEmail: '',
+            siteContactFirstName: '',
+            siteContactLastName: '',
+            siteContactPhone: '',
+            siteContactEmail: '',
+          }}
+          validationSchema={CreateSiteSchema}
+          onSubmit={(values) => {
+            handleSiteCreate({
```
Comment (fwpushan) [inline client/src/pages/private/SiteTable.js:102]:
```suggestion
            handleSiteCreate({...values, siteId: parseInt(values.siteId), allocation: parseInt(values.allocation)})
```
Is this mapping logic required?

File: server/routes/milestone-report.js [hunk 2]
```diff
@@ -8,21 +8,24 @@ const logger = require('../logger.js');
...
+  getReport,
+  getHiredParticipantsReport,
+  getHiredHealthRegionParticipantsReport,
+} = require('../services/reporting');
 
 // Router
 const router = express.Router();
 
-router.use(keycloak.allowRolesMiddleware('ministry_of_health'));
-router.use(keycloak.getUserInfoMiddleware());
```
Comment (fwpushan) [inline server/routes/milestone-report.js:17]: This is a common middleware for all nested routes.

File: server/routes/milestone-report.js [hunk 3]
```diff
@@ -58,4 +61,47 @@ router.get(
   })
 );
 
+router.get(
+  '/csv/hired/:regionId',
```
Comment (fwpushan) [inline server/routes/milestone-report.js:65]: No need for `regionId`, which is attached to the user.

File: server/routes/milestone-report.js [hunk 4]
```diff
@@ -8,21 +8,24 @@ const logger = require('../logger.js');
...
-router.use(keycloak.getUserInfoMiddleware());
-
 router.get(
   '/',
+  [keycloak.allowRolesMiddleware('ministry_of_health'), keycloak.getUserInfoMiddleware()],
   asyncMiddleware(async (req, res) => res.status(200).json({ data: await getReport() }))
 );
 
 router.get(
   '/csv/hired',
```
Comment (fwpushan) [inline server/routes/milestone-report.js:27]: We can reuse this route and service method.

File: server/tests/milestone-report.e2e.test.js [hunk 5]
```diff
@@ -31,4 +31,17 @@ describe('api-e2e test for route /api/v1/milestone-report', () => {
...
+    const header = await getKeycloakToken(healthAuthority);
+    const region = 'Vancouver Island';
+    const res = await request(app).get(`/api/v1/milestone-report/csv/hired/${region}`).set(header);
+    expect(res.status).toEqual(200);
+  });
+
+  it('should not let HA get general report', async () => {
+    const header = await getKeycloakToken(healthAuthority);
+    const res = await request(app).get(`/api/v1/milestone-report/csv/hired`).set(header);
+    expect(res.status).toEqual(500);
```
Comment (fwpushan) [inline server/tests/milestone-report.e2e.test.js:45]: the status will be 403 (Forbidden).

**Review by fwpushan — CHANGES_REQUESTED** [review]
We need some changes on the backend 
- We can reuse the existing route and service method. Almost identical service methods are triggering code duplication alerts in SonarScan.
- Fix tests

File: client/src/pages/private/SiteTable.js (hunk 1 above)
Comment (alva-fresh) [reply client/src/pages/private/SiteTable.js:102]: I didn't change anything here ☝️
I just moved it to a new component

File: server/routes/milestone-report.js (hunk 4 above)
Comment (alva-fresh) [reply server/routes/milestone-report.js:27]: I think we need a new endpoint

Comment (alva-fresh) [reply server/routes/milestone-report.js:61]: Actually, I'm curious - what would work best in this scenario?
- using the same endpoint for both HA and MoH (_assuming that validation done on FE_); or
- switch to different endpoints for different roles ?

File: client/src/pages/private/SiteTable.js (hunk 1 above)
Comment (fwpushan) [reply client/src/pages/private/SiteTable.js:102]: I am suggesting, just check that mapping logic has any merit or not

File: server/routes/milestone-report.js (hunk 4 above)
Comment (fwpushan) [reply server/routes/milestone-report.js:61]: if the underlying service method and processing is context-driven, then we can use the same route and service method.

**Review by fwpushan — APPROVED** [review]
LGTM 👍

**Pull request merged**